### PR TITLE
Generate notebook sources

### PR DESCRIPTION
**Pull request by karenfeng**
## What changes are proposed in this pull request?

Adds generated notebook sources for our documentation for easy diff-ability by converting from HTML to source code. Checks via CI that the generated sources are not out-of-date.

## How is this patch tested?
- [ ] Unit tests
- [x] Integration tests
- [x] Manual tests

To generate a notebook source yourself:
- Update your conda environment with the new dependencies
- Configure your Databricks CLI
- Run `gen-nb-src.py` with the input HTML

Example of CI catching failures: https://app.circleci.com/pipelines/github/projectglow/glow/1646/workflows/58e2deeb-7036-468e-8975-bcc3115a04a0/jobs/3878

### CODE REVIEW

File: docs/dev/gen-nb-src.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        for ext in SOURCE_EXTS:
+            source_path = os.path.join(SOURCE_DIR, rel_path + "." + ext)
+            if os.path.exists(source_path):
+                os.remove(source_path)
+                print("\tDeleted {}.".format(source_path))
+        return
+
+    print("Generating source file for {} under {} ...".format(html, SOURCE_DIR))
+
+    def run_cli_workspace_cmd(args):
```
Comment (henrydavidge) [inline docs/dev/gen-nb-src.py:41]: Why is this a nested definition?

File: docs/dev/gen-nb-src.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+SOURCE_EXTS = ['scala', 'py', 'r', 'sql']
+
+
+@click.command()
+@click.option('--html', required=True, help='Path of the HTML notebook.')
+@click.option('--cli-profile', default='docs-ci', help='Databricks CLI profile name.')
+@click.option('--workspace-tmp-dir', default='/tmp/glow-docs-ci', help='Temp workspace dir for import/export.')
+def main(html, cli_profile, workspace_tmp_dir):
+    assert os.path.commonpath([NOTEBOOK_DIR, html]) == NOTEBOOK_DIR, \
+        "HTML notebook must be under {} but got {}.".format(NOTEBOOK_DIR, html)
```
Comment (henrydavidge) [inline docs/dev/gen-nb-src.py:27]: nit: prefer f strings for interpolation

File: docs/dev/gen-nb-src.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    rel_path = os.path.splitext(os.path.relpath(html, NOTEBOOK_DIR))[0]
+
+    if not os.path.exists(html):  # html notebook was deleted
+        print("{} does not exist. Deleting the companion source file...".format(html))
+        for ext in SOURCE_EXTS:
+            source_path = os.path.join(SOURCE_DIR, rel_path + "." + ext)
+            if os.path.exists(source_path):
+                os.remove(source_path)
+                print("\tDeleted {}.".format(source_path))
```
Comment (henrydavidge) [inline docs/dev/gen-nb-src.py:36]: Why the tab?

File: docs/dev/gen-nb-src.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import uuid
+
+NOTEBOOK_DIR = 'docs/source/_static/notebooks'
+SOURCE_DIR = 'docs/source/_static/zzz_GENERATED_NOTEBOOK_SOURCE'
+SOURCE_EXTS = ['scala', 'py', 'r', 'sql']
+
+
+@click.command()
+@click.option('--html', required=True, help='Path of the HTML notebook.')
+@click.option('--cli-profile', default='docs-ci', help='Databricks CLI profile name.')
```
Comment (henrydavidge) [inline docs/dev/gen-nb-src.py:23]: Might it make more sense to rely on the databricks CLI's default profile?

File: docs/dev/gen-nb-src.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+@click.option('--workspace-tmp-dir', default='/tmp/glow-docs-ci', help='Temp workspace dir for import/export.')
```
Comment (henrydavidge) [inline docs/dev/gen-nb-src.py:24]: We should probably generate a temporary directory if one is not provided. In fact, maybe we should always just make a new one?

**Review by henrydavidge — COMMENTED** [review]
Had a few comments on the script. This is cool!

File: docs/dev/gen-nb-src.py (hunk 4 above)
Comment (karenfeng) [reply docs/dev/gen-nb-src.py:23]: I found this useful during manual testing, as I already have a default Databricks CLI profile and wanted to test this on the same shard as the CI.

File: docs/dev/gen-nb-src.py (hunk 5 above)
Comment (karenfeng) [reply docs/dev/gen-nb-src.py:24]: Actually, we do generate a temporary work directory under this; this serves as the root directory. I'll clarify in the description.

File: docs/dev/gen-nb-src.py (hunk 3 above)
Comment (karenfeng) [reply docs/dev/gen-nb-src.py:36]: We printed a line about the HTML file before this; here, we add a line for all associated generated source files (although reasonably, there should only be 1).

File: docs/dev/gen-nb-src.py (hunk 1 above)
Comment (karenfeng) [reply docs/dev/gen-nb-src.py:41]: This is just so we can leverage `cli_profile` from the main args, but I can split this out if it's clearer that way.

File: docs/dev/gen-nb-src.py (hunk 4 above)
Comment (henrydavidge) [reply docs/dev/gen-nb-src.py:23]: I mean that it makes sense to have this argument, but the default value should be the default profile. If you're running locally, I don't think there's any reason to expect a profile called `docs-ci` to exist. We can pass the `docs-ci` argument from our CI script.

File: docs/dev/gen-nb-src.py (hunk 1 above)
Comment (henrydavidge) [reply docs/dev/gen-nb-src.py:41]: I think that would be a bit cleaner.

File: python/environment.yml [hunk 6]
```diff
@@ -18,6 +18,8 @@ dependencies:
   - sphinx
   - sphinx_rtd_theme
   - pip:
+    - click==7.1.1 # Docs notebook source generation
```
Comment (henrydavidge) [inline python/environment.yml:21]: Can we install either of these from conda?

**Review by henrydavidge — COMMENTED** [review]
Few small comments

File: python/environment.yml [hunk 7]
```diff
@@ -1,20 +1,23 @@
...
-  - numpy=1.17.4
-  - scipy=1.4.1
-  - scikit-learn=0.22.1
   - nomkl # Skip MKL for local development
+  - numpy=1.17.4
   - pandas=0.25.3
   - pip
   - pyarrow=0.15.1
   - pyspark=3.0.0
+  - python=3.7
```
Comment (henrydavidge) [inline python/environment.yml:16]: I think it's actually standard to list python first since it is pretty important :).

**Review by henrydavidge — APPROVED** [review]
Nice!

**Pull request merged**